### ISSUE TITLE
Discussion on flying focal point

### ISSUE THREAD

**Issue by Zhentao-Liu** [CERN/TIGRE#621]
**Is your feature request related to a problem? Please describe.**
I recently process a CBCT scan with flying focal point, i.e., different focal points for odd/even frames. Actual focal point usually deviates from the ideal one, and such deviation could be described by "geo.offdetector". Although "geo.offdetector" could be input as an array to describe per frame focal point, but it reports bugs when running FDK algorithm. In FDK func of single_pass_algorithms.py, TIGRE currently performs zeropadding and perweighting2 first if "geo.offdetector" exsist, as shown in [code](https://github.com/CERN/TIGRE/blob/668f92e66b1e907155316f99bcd11a2a89afd8f4/Python/tigre/algorithms/single_pass_algorithms.py#L149). However, these two functions only accept 2-d vector input right now.

**Describe the solution you'd like**
Simple reimplementation is needed to make zeropadding and perweighting2 function accept array input.

**Additional context**
BTW, I wanna ask the unit of "geo.offdetector". I think it should be millimeter.

**Comment by AnderBiguri** [comment]
Apologies @Zhentao-Liu , this took much longer to fix than expected, I was very busy. 
May I ask again what the issue is? re-reading your text, its not very clear what you want. You say that geo.offDetector only accepts 2-d vector as input. What is the alternative you are looking for. What is what you want them to accept?

For your question in the end: the units are irrelevant, as long as you use the same units as in the rest of the geometry, such as DSD, geo.sVoxel etc.

**Comment by Zhentao-Liu** [comment]
Thanks for your reply. Maybe a little hard to solve this problem. Actually, I try to handle with curve-plane cbct scan with FFS. And I currently resort to [LEAP](https://github.com/LLNL/leap) for solution. It supports curve-plane geometry but do not support FFS. :<, pity. I temporally close this issue, and If I have any further questions I would re-open it again. Thanks anyway.

**Comment by AnderBiguri** [comment]
@Zhentao-Liu If you explain how FFS happens (I don't know) I may be able to explain how it may be done with TIGRE.

**Comment by Zhentao-Liu** [comment]
That would be great. Would you simply describe it. Thanks!

**Comment by AnderBiguri** [comment]
@Zhentao-Liu Sorry, maybe you misunderstood. I am asking for you to explain or give me references on how FFS works in commercial CT scanners, as I don't know. Otherwise I can't tell how to implement/describe it in TIGRE.

**Comment by Zhentao-Liu** [comment]
The basic idea is simply as follows. For even/odd frame, the focal spot is slightly different. Assume the physical center point is at [0, 0]pixel, for odd frame, the focal spot is at [0.25, 0]pixel for example, and for even frame, the focal spot is at [-0.25, 0]pixel for example. Then actually pixel spacing is half of the original physical pixel spacing, the minimal reconstruction voxel spacing could also be half, so FFS improve spaital resoution.

**Comment by AnderBiguri** [comment]
@Zhentao-Liu  

so, something like:

```
geo.offDetector = np.zeros((2,len(angles))
geo.offDetector[0,::2] =-0.25
geo.offDetector[0,1::2]=0.25
```

?

**Comment by Zhentao-Liu** [comment]
yes you are right. are there would be some bugs when running fdk directly.



---Original---
From: ***@***.***&gt;
Date: Sun, Jan 26, 2025 23:37 PM
To: ***@***.***&gt;;
Cc: "Zhentao ***@***.******@***.***&gt;;
Subject: Re: [CERN/TIGRE] Discussion on flying focal point (Issue #621)




 
@Zhentao-Liu
 
so, something like:
 geo.offDetector = np.zeros((2,len(angles)) geo.offDetector[0,::2] =-0.25 geo.offDetector[0,1::2]=0.25  
?

**Comment by AnderBiguri** [comment]
Are there? try it and do report if you find any please!
